### PR TITLE
coprocessor: Add plus RPN functions for Int, Real and Decimal

### PR DESCRIPTION
**Pull request by sticnarf**
## What have you changed? (mandatory)

This PR adds `plus` RPN functions for Int, Real and Decimal. Code and tests are mainly adapted from `builtin_arithmetic.rs`.

The `map_int_sig` function is extracted from the original `map_compare_int_sig` to support different mappings.

## What are the type of the changes? (mandatory)

Improvement.

## How has this PR been tested? (mandatory)

Unit test

## Does this PR affect documentation (docs) or release note? (mandatory)

No
## Does this PR affect tidb-ansible update? (mandatory)

No

### CODE REVIEW

File: src/coprocessor/dag/rpn_expr/impl_arithmetic.rs [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+    }
+}
+
+impl<A: ArithmeticOp> Clone for RpnFnArithmetic<A> {
+    fn clone(&self) -> Self {
+        Self::new()
+    }
+}
+
+impl<A: ArithmeticOp> Copy for RpnFnArithmetic<A> {}
```
Comment (lonng) [inline src/coprocessor/dag/rpn_expr/impl_arithmetic.rs]: Why not derive `Copy` directly?

Comment (sticnarf) [reply src/coprocessor/dag/rpn_expr/impl_arithmetic.rs]: A rust bug https://github.com/rust-lang/rust/issues/26925 ...

File: src/coprocessor/dag/rpn_expr/impl_arithmetic.rs [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+                .push_param_with_field_type(rhs, rhs_field_type)
+                .evaluate::<Int>(sig)
+                .unwrap();
+            assert_eq!(output, expected, "{:?}, {:?}", output, expected);
+        }
+    }
+
+    #[test]
+    fn test_arithmetic_real() {
+        let test_cases = vec![(PlusReal, Some(1.01001), Some(-0.01), Some(1.00001))];
```
Comment (rleungx) [inline src/coprocessor/dag/rpn_expr/impl_arithmetic.rs]: I think we also need to test the case which will overflow.

Comment (sticnarf) [reply src/coprocessor/dag/rpn_expr/impl_arithmetic.rs]: Real overflow case covered.

File: src/coprocessor/dag/rpn_expr/impl_arithmetic.rs [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+pub struct UintIntPlus;
+
+impl ArithmeticOp for UintIntPlus {
+    type T = Int;
+
+    fn calc(lhs: &Int, rhs: &Int) -> Result<Option<Int>> {
+        IntUintPlus::calc(rhs, lhs)
+    }
+}
+#[derive(Debug)]
```
Comment (breeswish) [inline src/coprocessor/dag/rpn_expr/impl_arithmetic.rs]: New line before this one.

File: src/coprocessor/dag/rpn_expr/impl_arithmetic.rs [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+                .evaluate::<Int>(sig)
+                .unwrap();
+            assert_eq!(output, expected, "{:?}, {:?}", output, expected);
+        }
+    }
+
+    #[test]
+    fn test_arithmetic_real() {
+        let test_cases = vec![
+            (PlusReal, Some(1.01001), Some(-0.01), Some(1.00001), false),
```
Comment (breeswish) [inline src/coprocessor/dag/rpn_expr/impl_arithmetic.rs]: Looks like we can simply move the sig into `.evaluate(..)`.

File: src/coprocessor/dag/rpn_expr/impl_arithmetic.rs [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+            (PlusReal, Some(1e308), Some(1e308), None, true),
+        ];
+        for (sig, lhs, rhs, expected, is_err) in test_cases {
+            let output = RpnFnScalarEvaluator::new()
+                .push_param(lhs)
+                .push_param(rhs)
+                .evaluate::<Real>(sig);
```
Comment (breeswish) [inline src/coprocessor/dag/rpn_expr/impl_arithmetic.rs]: I guess Real can be auto inferred

**Comment by sticnarf** [comment]
> Great job! Honestly I would like to see #4729 being merged first so that I don't need to fix conflicts due to newly introduced f64 in this PR XD

OK. I will update this PR once #4729 is merged.

**Comment by breeswish** [comment]
@AndreMouche This PR need to be updated..
